posts/Gunnar/2019-01-23-testing-cdi-beans-and-persistence-layer-under-java-se.adoc

```diff
@@ -0,0 +1,437 @@
+
+[source,java]
+----
+@ApplicationScoped
+public class EntityManagerFactoryProducer {
+
+    @Inject
+    private BeanManager beanManager;
+
+    @Produces
+    @ApplicationScoped
+    public EntityManagerFactory produceEntityManagerFactory() {
+        Map<String, Object> props = new HashMap<>();
+        props.put("javax.persistence.bean.manager", beanManager);
+        return Persistence.createEntityManagerFactory("myPu", props);
+    }
+
+    public void close(@Disposes EntityManagerFactory entityManagerFactory) {
+        entityManagerFactory.close();
+    }
+}
+----
+
+This lets us make use of dependency injection within JPA entity listeners:
+
+[source,java]
+----
+@ApplicationScoped
+public class SomeListener {
+
+    private final GreetingService greetingService;
+
+    @Inject
+    public SomeListener(GreetingService greetingService) {
+        this.greetingService = greetingService;
+    }
+
+    @PostPersist public void onPostPersist(TestEntity entity) {
+        greetingService.greet(entity.getName());
+    }
+}
+----
+
+== Declarative Transaction Control via `@Transactional` and transactional event observers
+
+The last missing piece to satisfy our original requirements is support for the `@Transactional` annotation and transactional event observers.
+This one is a bit more complex to tackle, as it requires the integration of a transaction manager compatible with JTA (the Java Transaction API).
+
+In the following we're going to use http://narayana.io/[Narayana] which also is the transaction manager used in WildFly.
+For Narayana to work, a JNDI server is needed, from which it can obtain the JTA datasource.
+Furthermore, the Weld JTA module is required.
+Please refer to the https://github.com/hibernate/hibernate-demos/blob/master/other/cdi-jpa-testing/pom.xml[pom.xml] of the example project for the exact artifact ids and versions.
+
+With these dependencies in place, the next step is to plug in a custom `ConnectionProvider` into Hibernate ORM,
+which makes sure that Hibernate ORM works with `Connection` objects that use transactions managed by Narayana.
+Thankfully, my colleague Gytis Trikleris has provided https://github.com/jbosstm/quickstart/blob/master/jta-and-hibernate-standalone/src/main/java/org/jboss/narayana/quickstart/jta/TransactionalConnectionProvider.java[such implementation] already as part of the Narayana examples on GitHub.
+I'm shamelessly going to copy this implementation:
+
+[source,java]
+----
+public class TransactionalConnectionProvider implements ConnectionProvider {
+
+    public static final String DATASOURCE_JNDI = "java:testDS";
+    public static final String USERNAME = "sa";
+    public static final String PASSWORD = "";
+
+    private final TransactionalDriver transactionalDriver;
+
+    public TransactionalConnectionProvider() {
+        transactionalDriver = new TransactionalDriver();
+    }
+
+    public static void bindDataSource() {
```

Comment:
How is this called? With some reflection magic in Narayana?

Reply:
Done.